build_runner_core/lib/src/package_graph/package_graph.dart

```diff
@@ -94,11 +115,16 @@ class PackageGraph {
     final consistentlyOrderedPackages = packageConfig.packages.toList()
       ..sort((a, b) => a.name.compareTo(b.name));
     for (final package in consistentlyOrderedPackages) {
-      var isRoot = package.name == rootPackageName;
+      // The pubspec.lock file does not contain the workspace packages.
+      // We treat them as "path" packages in terms of mutability.
+      final isWorkspacePackage = !dependencyTypes.containsKey(package.name);
+      final isRoot = package.name == rootPackageName;
       nodes[package.name] = PackageNode(
           package.name,
           package.root.toFilePath(),
-          isRoot ? DependencyType.path : dependencyTypes[package.name],
+          isWorkspacePackage
```

Comment:
What if we just did `dependencyTypes[package.name] ?? DependencyType.path` instead here, and remove `isWorkspacePackage`? That just feels a bit better even though the logic is identical, we wouldn't as scary of a comment. Instead it can just be `if the package is missing from pubspec.lock, assume it is a path dependency`.

Reply:
Done